[CDAP-13985] UI stack upgrade - React and associated libraries upgrade

ajainarayanan opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-13985
Build: https://builds.cask.co/browse/CDAP-UDUT23

**TL;DR-**
- Upgrades react and react-dom to 16.4.1
- The purpose is,
    - To be able to use new features (patterns) added as part of the new design.
    - To get perf improvements added as part of the new rewrite in react.
    - To be able to upgrade to newer version of the libraries/components that we use for new features/fixes/perf improvements
- Some new features causes UI to not behave the same way it used to. Fixed most of the cases I saw. This involves fixing
  - keys in list elements
  - passing `undefined` for event handlers instead of null
  - fixing proptypes
- Changed the way we use dlls for dev vs prod. Now in dev build we use dev dll files to be able debug or get stack trace when error happens during development (mainly keys and proptypes).

**Longer version**
- React 15.x -> 16.x was a complete rewrite of react
  - This involves perf improvements on existing functionality
  - Newer features that can make life so much better
- New features that we are directly/indirectly using
  - [Portals](https://reactjs.org/docs/portals.html) (modals)
  - [Context](https://reactjs.org/docs/context.html) (react-redux underneath no longer uses the old context api)
  - [Fragments](https://reactjs.org/docs/fragments.html) (We don't use it directly now but we will very soon)
  - [Error boundaries](https://reactjs.org/docs/error-boundaries.html) (Very basic usage. We need to be able leverage this better to surface 500 js errors and 404 redirects)
  - [Create ref](https://reactjs.org/blog/2018/03/29/react-v-16-3.html#createref-api) (We don't use this feature yet but we need to modify our usage of `ref`s to the new way.)

**Breaking changes**
-  **_ErrorBoundaries_**: Any JS error that happen in the render function will cause react to unmount the entire app. The reasoning is its better to unmount the entire app than showing something that is corrupted/unstable state. Today we have `ErrorBoundary` component that will handle this case. But its not intelligent enough to catch all. The important part is `ErrorBoundary` component will ONLY catch errors that happen in `render` functions of the children ([Error Boundary criteria](https://reactjs.org/docs/error-boundaries.html#introducing-error-boundaries)).

- **_Portals_**: Since modals switched to using portals placing the modal component (say AbstractWizard) inside the button tag will make the modal close right away as the "button" is treated as the parent of modal and all click events that happen in modal is eventually surfaced to parent which handles it as a click on the button and closes the modal. So we need to separate the button that opens the modal and the actual modal component into two different tags(divs)

- **_Lifecycle methods deprecation_** [Reference](https://reactjs.org/blog/2018/03/29/react-v-16-3.html#component-lifecycle-changes) We will eventually should do this to be able to use react's [Async rendering](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html)

**NOTE**:
- The major addition in this PR is because of the new dev dll generated to be used during development.
- There are still some css changes required. This is because of bootstrap patch. Bootstrap 4 defined global font and header sizes to be in rems and we don't maintain the same font size between react and angular. Hence the discrepancies. I didn't address all of them now as it required changing the root (global) styles.

ajainarayanan committed:
[CDAP-13985] UI stack upgrade - React and associated libraries upgrade

  - Unifies build processes to generate appropriate build files based on mode (dev vs production)
  - Upgrades react to 16.4 + libraries that depend on react
  - Modifies Gulpfile to fetch react from right path based on react & react-dom upgrade
  - Fixes RulesEngine for react errors
  - Fixes FieldLevelLineage for react errors
  - Fixes DataPrep for react errors
  - Fixes MMDS(Experiments) for react errors
  - Fixes AppDetailedView, DatasetStreamCards & StreamDetailedView for react errors
  - Fixes PlusButton & PlusButtonModal for react errorss
  - Fixes CaskWizards for react errors
  - Fixes BreadCrumb, GroupedBarChart, SpotlightModal, Tags & VegaLiteChart for react errors
  - Fixes DSVEditor to pass in right initial state for its store
  - Fixes FastActionButton, StartStopAction & SetPreferenceModal for react errors
  - Fixes WizardModal for react errors
  - Fixes Popover for react errors
  - Fixes Overview for react errors
  - Fixes Card to properly handle onClick on card (if within the card do nothing)
  - Fixes MarketPlaceEntity for react errors
  - Fixes Header for react errors
  - Fixes ConfigurableTab for react errors
  - Fixes DeleteEntityBtn for react errors
  - Fixes NamespaceDropdown for react errors
  - Fixes DataPrepConnections to pass in right initial state to the store
  - Fixes HydratorModal for react errors
  - Fixes Profile create for bootstrap errors
  - Fixes ColumnActionsDropdown to adjust width because of reacstrap upgrade
  - Fixes search container in entity list view to use new bootstrap markup for input group
  - Fixes Header dropdown after reskinning the header in 5.0
  - Fixes ResourceCenter to not close on clicking anywhere in the modal
  - Fixes error boundary to handle a 500 error appropriately
  - Fixes 500 error page to be a tad better
  - Fixes AddConnection popover to use the right popover
  - Fixes connections popover to do crud operations to pass in the right props
  - Fixes the UnControlledPopover to accept right props based on reactstrap upgrade
  - Fixes WorkspaceTabs popover to pass in the right props
  - Converts CopyableId to use Popover instead of UnControlledTooltip from reactstrap
  - Adds basic error boundary support
  - Adds bootstrap 4 patch styles to make sure using bootstrap4-alpha on some react based libraries doesn't break UI
  - Updates all html templates to have a class on body to add extra namespace for bootstrap patch styles
  - NOT PROPER WORKING cut directive. The popovers are not positioned correctlyr
  - Fixes the webpack configs and cdap and login html files to use mode to properly include the appropriate dll files
  - Fixes KeyvaluePairs component to subscrie to store in componentDidMount instead of constructor
  - [CDAP-13985] Upgrade Babel + Webpack & add the ability to write typescript for UI development

    - Updates Babel, webpack node modules + adds typescript, tslint & @types/* node modules for typescript integration
    - Updates webpack configurations based on upgrade to babel and typescript
    - Adds a sample typescript component to be rendered in UI
    - Updates dlls post webpack upgrade

  - Fixes references of CreateStreamWithUploadWizardConfig
  - Fixes the way we show tooltips + adds bootstrap patch for tooltip styling for bootstrap 4

tonybach commented:
Did a first pass through the UI. In general things look good, I just found some weird issues with the Add Namespace modal.

![namespace-add-problems](https://user-images.githubusercontent.com/6516002/43493401-bfecd440-94e2-11e8-929c-d5dc578cb31f.gif)

The key-value pairs not erasing even when we close the modal is a known issue, but there seems to be something wrong with the KeyValuePairs after the changes, at least when used in this wizard.

tonybach commented:
In React the label for the 'Add entity' modal doesn't seem centered, and in Angular it becomes really small.

![modal-header](https://user-images.githubusercontent.com/6516002/43493919-29c54d0a-94e5-11e8-98a3-de043e0391ae.gif)

ajainarayanan commented:
Noticed. 
<img width="341" alt="screen shot 2018-08-01 at 7 36 44 am" src="https://user-images.githubusercontent.com/1452845/43556867-d31d7036-95b6-11e8-9415-147ca3ec3d8c.png">

ajainarayanan commented:
We need to fix all popovers and tooltips used from `reactstrap` to be replaced with inhouse `Popover` component (because of change in tether usage by reactstrap)

tonybach commented:
Found a couple of other bugs:
1. Clicking on 'Insights' in DataPrep just shows a spinning wheel forever. In the console, I get the error `Cannot read property 'oneOfType' of undefined`.

2. There's a gap between the namespace dropdown and the navbar, which didn't exist before.

![image](https://user-images.githubusercontent.com/6516002/43733703-631c6aa0-996a-11e8-94bf-5de0da4a6826.png)

3. Clicking on or hovering on the (i) icons in this 'Ingest Data' modal doesn't show any tooltip.

![image](https://user-images.githubusercontent.com/6516002/43733764-8f90045c-996a-11e8-978e-65fa1fbdc503.png)

4. Found a weird issue with the 'Copy to Clipboard' tooltip in the 'Directives' modal in Analytics.

![tooltip-on-popover](https://user-images.githubusercontent.com/6516002/43733822-b773945c-996a-11e8-890a-3c2a600ac705.gif)

elfenheart commented on cdap-ui/app/cdap/components/KeyValuePairs/index.js:
```diff
@@ -92,22 +92,26 @@ let KeyValuePairCopy = connect(
       type: KeyValueStoreActions.onUpdate,
       payload: {pairs: keyValues.pairs}
     });
+    console.log('Inside keyvalue pairs constructor');
```
remove

tonybach commented:
Need to fix this, otherwise the user won't be able to click 'Play' on any program in Overview 

![error-on-fastaction-hover](https://user-images.githubusercontent.com/6516002/43857452-33e09a80-9b00-11e8-8b1d-29beff48c277.gif)

ajainarayanan commented:
Have addressed all the issues mentioned above. Things not addressed in this PR are,
Copy to clipboard - https://issues.cask.co/browse/CDAP-14066
Weird gap in namespace dropdown - https://issues.cask.co/browse/CDAP-14067

ajainarayanan commented:
Will open separate PRs for the referenced JIRAs. Merging this for now.
